### PR TITLE
Codelist/code usage

### PR DESCRIPTION
**Pull request by luciecastella**
This PR is constructed on top of PR #177 

Code now have attributes: name, description and attributes
VariableCode was created as a subclass of Code, with additional attributes: unit, weight, region_aggregation,...

A CodeList's mapping is now a Dict[str, Code]. 

Currently, 7 tests are still failing because of the change in CodeList call. But I will still work on it in the next days.

### CODE REVIEW

**Review by phackstock — COMMENTED** [review]
Looks quite good to me overall.
The one thing that could be improved (although that might be a bit much for this PR) is the way the tags are handled right now. It's probably a little bit overly complicated where the codes are initially created and then re-created by `VariableCode._replace_tags`. In my opinion, there should ideally not be the need to override the `Code._replace_tags` method. I think the most elegant way would be to include the tag replacement in the `Code` constructor.
I wouldn't worry about it too much though as the primary focus should be to get the tests working and then get the first version of the specialized `Code` classes merged. I think I'll just tackle the clean up in a follow up PR then.

**Review by phackstock — COMMENTED** [review]
Just two small comments that I forgot earlier.

File: nomenclature/definition.py [hunk 1]
```diff
@@ -93,8 +93,8 @@ def check_aggregate(self, df: IamDataFrame, **kwargs) -> None:
         with adjust_log_level(level="WARNING"):
             for code in df.variable:
                 attr = self.variable[code]
-                if attr.get("check-aggregate", False):
-                    components = attr.get("components", None)
+                if attr.dict().get("check-aggregate", False):
```
[the remark this answers is not in the document]
Comment (luciecastella) [reply nomenclature/definition.py]: Smart thank you

**Comment by luciecastella** [comment]
Ooook I think that's it. Hope I did not forget anything. 
@phackstock I did not have time to tackle what you mentioned with `_replace_tags` so yes I will let you do it in another PR.
This is a longer PR, I hope it meets what you were expecting. I'll leave it to you from there.

**Comment by phackstock** [comment]
Closing in favor of #188 as there are some changes necessary and I do not have write permissions on the branch in this PR.

**Pull request closed without merging**